e2e/nomostest/nt.go

```diff
@@ -793,59 +788,69 @@ func (nt *NT) MustDeleteGatekeeperTestData(file, name string) {
 	nt.MustKubectl("delete", "-f", absPath, "--ignore-not-found", "--wait")
 }
 
-// PortForwardOtelCollector forwards the otel-collector pod.
-func (nt *NT) PortForwardOtelCollector() {
-	// Retry otel-collector port-forwarding in case it is in the process of upgrade.
-	took, err := Retry(60*time.Second, func() error {
-		pod, err := nt.GetDeploymentPod(ocmetrics.OtelCollectorName, ocmetrics.MonitoringNamespace)
-		if err != nil {
-			return err
-		}
-		// The otel-collector forwarding port needs to be updated after otel-collector restarts or starts for the first time.
-		// It sets otelCollectorPodName and otelCollectorPort to point to the current running pod that forwards the port.
-		if pod.Name != nt.otelCollectorPodName {
-			port, err := nt.ForwardToFreePort(ocmetrics.MonitoringNamespace, pod.Name, testmetrics.MetricsPort)
-			if err != nil {
-				return err
-			}
-			nt.otelCollectorPort = port
-			nt.otelCollectorPodName = pod.Name
+// portForwardOtelCollector forwards the otel-collector pod.
+func (nt *NT) portForwardOtelCollector() {
+	if nt.otelCollectorPort == 0 {
+		var mux sync.Mutex
+		setOtelCollectorPort := func(localPort int, podName string) {
+			mux.Lock()
+			defer mux.Unlock()
+			nt.otelCollectorPort = localPort
 		}
-		return nil
-	})
-	if err != nil {
-		nt.T.Fatal(err)
+		ctx, cancel := context.WithCancel(nt.Context)
+		nt.T.Cleanup(cancel)
+		nt.keepPortForwardOpen(ctx, ocmetrics.MonitoringNamespace, ocmetrics.OtelCollectorName, testmetrics.MetricsPort, setOtelCollectorPort)
+
 	}
-	nt.T.Logf("took %v to wait for otel-collector port-forward", took)
 }
 
-// PortForwardGitServer forwards the git-server deployment to a port.
-func (nt *NT) PortForwardGitServer() {
+// portForwardGitServer forwards the git-server deployment to a port.
+func (nt *NT) portForwardGitServer() {
 	nt.T.Helper()
 
-	pod, err := nt.GetDeploymentPod(testGitServer, testGitNamespace)
-	if err != nil {
-		nt.T.Fatal(err)
-	}
-	podName := pod.Name
-
-	if podName != nt.gitRepoPodName {
-		port, err := nt.ForwardToFreePort(testGitNamespace, podName, ":22")
-		if err != nil {
-			nt.T.Fatal(err)
+	if nt.gitRepoPort == 0 {
+		var mux sync.Mutex
+		prevPodName := ""
+		setGitRepoPort := func(localPort int, podName string) {
+			mux.Lock()
+			defer mux.Unlock()
+			nt.T.Logf("updating port-forward %s:%d -> %s:%d", prevPodName, nt.gitRepoPort, podName, localPort)
+			// pod unchanged, don't reset
+			if prevPodName == "" || prevPodName == podName {
+				nt.gitRepoPort = localPort
+				prevPodName = podName
+				return
+			}
+			// allRepos specifies the slice all repos for port forwarding.
+			var allRepos []types.NamespacedName
+			for repo := range nt.RootRepos {
+				allRepos = append(allRepos, RootSyncNN(repo))
+			}
+			for repo := range nt.NonRootRepos {
+				allRepos = append(allRepos, repo)
+			}
+			// re-init all repos
+			InitGitRepos(nt, allRepos...)
```

Comment:
I don't understand why all this logic is in the setGitRepoPort function...?

Reply:
> The git server also needs to have its git repositories re-initialized because it uses emptyDir storage.

We could alternatively use a PersistentVolume but this seemed like a simpler solution

Reply:
Do they need to be re-initialized ever time port-forwarding restarts? Or just every time it's a new pod? Or are those always the same?

Reply:
yeah this only gets reached when the pod restarts. There's also a check above to see if the pod name changed